onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_api_helper.cc

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "nnapi_api_helper.h"
+
+#include "core/common/inlined_containers_fwd.h"
+#include "core/providers/nnapi/nnapi_builtin/builders/model_builder.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_lib/nnapi_implementation.h"
+#include "core/common/logging/logging.h"
+
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
+
+namespace onnxruntime {
+namespace nnapi {
+
+/**  How feature level works for NNAPI. refer to https://developer.android.com/ndk/reference/group/neural-networks
+ *
+ * NNAPI device feature level is closely related to NNAPI runtime feature level
+    (ANeuralNetworks_getRuntimeFeatureLevel), which indicates an NNAPI runtime feature level
+    (the most advanced NNAPI specification and features that the runtime implements).
+    An NNAPI device feature level is always less than or equal to the runtime feature level.
+ *
+ * On Android devices with API level 30 and older, the Android API level of the Android device
+    must be used for NNAPI runtime feature discovery.
+    Enum values in FeatureLevelCode from feature level 1 to 5 have their
+    corresponding Android API levels listed in their documentation,
+    and each such enum value equals the corresponding API level.
+    This allows using the Android API level as the feature level.
+    This mapping between enum value and Android API level does not exist for
+    feature levels after NNAPI feature level 5 and API levels after S (31).
+
+ */
+static int32_t GetNNAPIRuntimeFeatureLevel(const NnApi& nnapi_handle) {
+  int32_t runtime_level = static_cast<int32_t>(nnapi_handle.nnapi_runtime_feature_level);
+
+#ifdef __ANDROID__
+  int device_api_level = android_get_device_api_level();
+  runtime_level = (device_api_level < __ANDROID_API_S__) ? device_api_level : runtime_level;
+#endif
+  return runtime_level;
+}
+
+/**
+ * Get the max feature level supported by all target devices.
```

Comment:
this is what I'd expect this wording to mean:
"max feature level supported by all target devices" = `min(device feature level for each target device)`
"max feature level supported by any target device" = `max(device feature level for each target device)`

as "device feature level" is the max supported feature level for a particular device

Reply:
suggested wording (also in other places):
"get the max feature level supported by all target devices" -> "get the max feature level supported by target devices"

the "all" implies that the returned feature level is supported by all devices, but it is actually the maximum feature level across all devices.